extmod/moddeflate.c: Add deflate.DeflateIO providing compression and decompression.

jimmo opened the pull request:
This replaces https://github.com/micropython/micropython/pull/11879 -- because there was no way to provide a `gzip.GzipFile` that exactly matched CPython's (e.g. because we _need_ the `wbits` parameter), instead we will put this functionality in a separate module/class. I have updated https://github.com/micropython/micropython-lib/pull/694 to also provide a CPython-compatible `gzip` module.

Updated copy of the https://github.com/micropython/micropython/pull/11879 description below:

---

Summary: Adds gzip/zlib/raw-deflate compression support. (See issues #7228, #5590, and earlier PRs #8195, #5613).

Historically MicroPython provided a CPython-compatible implementation of `zlib.decompress` as well as a MicroPython-specific `zlib.DecompIO` which is similar to CPython's `zlib.decompressobj` except more useful as a stream wrapper, for example to decompress a file or socket, because you pull (read) decompressed data out of it, rather than pushing (write) compressed data into it.

However, CPython also provides `gzip.GzipFile` which works exactly like MicroPython's `DecompIO`, the only limitation is that it only supports gzip-format data (whereas `DecompIO` supported the same `wbits` argument that `zlib.decompress` supported).

This PR removes the `zlib` module, and replaces it with a `deflate` module providing a stream wrapper class `DeflateIO` that is essentially DeflateIO with compression support, but also suitable as a building block for implementing `gzip.GzipFile`. Because it no longer part of the `zlib` module, it also has a simpler API for specifying format and window size.

For backwards compatibility, I have written a pure-Python implementation of the functionality of the former `zlib` module (including now `zlib.compress`), as well as the `gzip` module (including `compress`, `decompress`, `open`, and `GzipFile`), which will be published to micropython-lib (https://github.com/micropython/micropython-lib/pull/694) and can be optionally installed via mip (or frozen). Even though they are very small (~450 bytes each), they are not frozen by default, with the justification:
- (De)compression isn't a feature that everyone uses.
- Even in applications that use (de)compression, DeflateIO provides a much more memory-efficient way to work with reading/writing compressed data to/from a stream (which is the typical use case).
- GzipFile not supporting the wbits argument makes it awkward to use in a general-purpose way.

This PR includes the following changes:
- Add a shared `mp_stream___exit___obj` that avoids each file/stream-like object needing to implement this themselves. (This is a code size optimisation saving ~200 bytes)
- Add a memory-efficient implementation of LZ77 compression to lib/uzlib (Written by @dpgeorge ).
- General code size optimisations for lib/uzlib. (Saves around 650 bytes)
- Code cleanup and simplication for lib/uzlib.
- Remove the `zlib` module.
- Add the `deflate` module.
- Add tests for `deflate.DeflateIO`.
- Update docs for `zlib` and add docs for `gzip` to describe the micropython-lib versions.
- Add docs for `deflate`.

Overall this PR adds +1156 bytes to PYBV11 (i.e. adding compression support). This is made up of -2840 (remove zlib), +2928 (add decompression-only DeflateIO), +1068 (add compression support to DeflateIO).

Currently compression support (via `MICROPY_PY_DEFLATE_COMPRESS`) is enabled at the same level ("extra features") as decompression (via `MICROPY_PY_DEFLATE`), but I think we should consider making it either "full features" or "everything". I do think we should enable it on Unix/Windows though (but really I think that means we should make it "full" and move Unix/Windows to "full").

Compared to #8195 & #5613 which are quite similar (cc @harbaum, @andrewleech) this PR adds support for streaming as well as non-gzip (i.e. zlib/raw) formats and a code size reduction as well as reduced memory usage in the compressor (see https://github.com/micropython/micropython/pull/11879#issuecomment-1608959558 for more notes from @dpgeorge).

_This work was funded through GitHub Sponsors, by specific request from a supporter._

dpgeorge committed:
lib/uzlib: Add memory-efficient, streaming LZ77 compression support.

The compression algorithm implemented in this commit uses much less memory
compared to the standard way of implementing it using a hash table and
large look-back window.  In particular the algorithm here doesn't allocate
hash table to store indices into the history of the previously seen text.
Instead it simply does a brute-force-search of the history text to find a
match for the compressor.  This is slower (linear search vs hash table
lookup) but with a small enough history (eg 512 bytes) it's not that slow.
And a small history does not impact the compression too much.

To give some more concrete numbers comparing memory use between the
approaches:

- Standard approach: inplace compression, all text to compress must be in
  RAM (or at least memory addressable), and then an additional 16k bytes
  RAM of hash table pointers, pointing into the text

- The approach in this commit: streaming compression, only a limited amount
  of previous text must be in RAM (user selectable, defaults to 512 bytes).

To compress, say, 1k of data, the standard approach requires all that data
to be in RAM, plus an additional 16k of RAM for the hash table pointers.
With this commit, you only need the 1k of data in RAM.  Or if it's
streaming from a file (or elsewhere), you could get away with only 256
bytes of RAM for the sliding history and still get very decent compression.

In summary: because compression takes such a large amount of RAM (in the
standard algorithm) and it's not really suitable for microcontrollers, the
approach taken in this commit is to minimise RAM usage as much as possible,
and still have acceptable performance (speed and compression ratio).

Signed-off-by: Damien George <damien@micropython.org>

jimmo committed:
lib/uzlib/defl_static: Implement some code size improvements.

This commit makes the following changes:
- Replace 256-byte reverse-bits-in-byte lookup table with computation.
- Replace length and distance code lookup tables with computation.
- Remove comp_disabled check (it's unused).
- Make the dest_write_cb take the data pointer directly, rather than the
  Outbuf.

Saves 500 bytes on PYBV11.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
extmod/moddeflate: Add deflate module providing the DeflateIO class.

This provides similar functionality to the former zlib.DecompIO and
especially CPython's gzip.GzipFile for both compression and decompression.

This class can be used directly, and also can be used from Python to
implement (via io.BytesIO) zlib.decompress and zlib.compress, as well as
gzip.GzipFile.

Enable/disable this on all ports/boards that zlib was previously configured
for.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
py/stream: Add mp_stream___exit___obj that calls mp_stream_close.

There are enough places that implement __exit__ by forwarding directly to
mp_stream_close that this saves code size.

For the cases where __exit__ is a no-op, additionally make their
MP_STREAM_CLOSE ioctl handled as a no-op.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
all: Remove the zlib module.

This will be replaced with a new deflate module providing the same
functionality, with an optional frozen Python wrapper providing a
replacement zlib module.

binascii.crc32 is temporarily disabled.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib: Clean up tinf -> uzlib rename.

This library used a mix of "tinf" and "uzlib" to refer to itself.  Remove
all use of "tinf" in the public API.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib: Combine zlib/gzip header parsing to allow auto-detect.

This supports `wbits` values between +40 to +47.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs/library/deflate: Add docs for deflate.DeflateIO.

Also update zlib & gzip docs to describe the micropython-lib modules.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib/tinflate: Implement more compact lookup tables.

Saves 68 bytes on PYBV11.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib/defl_static: Optimize zlib_start/finish_block.

Collapsing the two adjacent calls to outbits saves 32 bytes.

Bringing defl_static.c into lz77.c allows better inlining, saves 24 bytes.

Merge the Outbuf/uzlib_lz77_state_t structs, a minor simplification that
doesn't change code size.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib/lz77: Always use separate history buffer.

Because we only use the streaming source, this is just extra code size.

Saves 64 bytes on PYBV11.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
lib/uzlib: Add a source_read_data var to pass to source_read_cb.

For better abstraction for users of this API.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
tests/extmod: Add deflate.DeflateIO tests.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Some quick performance tests:
On PYBV11, decompression is unchanged 781ms (from 783ms) to decompress a 7.4kiB gzip file (into 14kiB) 25 times (using DeflateIO vs DecompIO from BytesIO using readinto to avoid allocations). 
On unix, decompression on MicroPython (DeflateIO) is roughly 7 times slower than in Python (gzip.GzipFile). 465ms vs 65ms for 1000 iterations of the same test.

For compression (on the unix port), MicroPython (wbits=8) is about 8x slower than Python (wbits=15) to compress the same file. (4x if you use wbits=6).

In terms of bytes/second:
On Unix, CPython compresses at about 45MiB/s, whereas MicroPython at about 2.8MiB/s. CPython decompresses at about 110MiB/s, whereas MicroPython is about 18MiB/s. 
On PYBV11, MicroPython compresses at 63kiB/s and decompresses at 245kiB/s.

MicroPython's decompression performance is impacted slightly by not buffering its input as it reads byte at a time from the underlying stream. On PYBV11 it can be increased to 280kiB/s by adding a 4 byte buffer. Diminishing marginal returns for a bigger buffer. Similar relative gain on Unix (22MiB/s).

MicroPython's compression performance is limited by the linear searching in the history window (it's not meant to be fast, it's meant to use less memory).

andrewleech commented:
I'm really excited to see this coming together, thanks @jimmo !

dpgeorge commented:
Re performance: compiling relevant `lib/uzlib` files with `-O2` will probably go a long way to improve performance.

dpgeorge commented on extmod/moddeflate.c:
```diff
@@ -0,0 +1,411 @@
+        // -----CMF------  ----------FLG---------------
+        // CINFO(5) CM(3)  FLEVEL(2) FDICT(1) FCHECK(5)
+        uint8_t buf[] = { 0x08, 0x80 };     // CM=2 (deflate), FLEVEL=2 (default), FDICT=0 (no dictionary)
+        buf[0] |= MAX(self->window_bits - 8, 1) << 4;     // base-2 logarithm of the LZ77 window size, minus eight.
```
Should this use `wbits`?  In case `self->window_bits==0`?

jimmo replied:
Good catch. Fixed.

dpgeorge commented on extmod/moddeflate.c:
```diff
@@ -0,0 +1,411 @@
+        { MP_QSTR_close, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_rom_obj = MP_ROM_FALSE} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
```
Possible make these positional-only args, to reduce code size and allow a natmod version.

jimmo replied:
Done.

dpgeorge commented on extmod/moddeflate.c:
```diff
@@ -0,0 +1,411 @@
+STATIC mp_uint_t deflateio_write(mp_obj_t self_in, const void *buf, mp_uint_t size, int *errcode) {
+    mp_obj_deflateio_t *self = MP_OBJ_TO_PTR(self_in);
+
+    if (self->stream == MP_OBJ_NULL || !deflateio_init_write(self)) {
```
I swapped the ordering here of these checks, to check if stream is null first (else it would crash if a write was done after a close).

dpgeorge commented:
I added a new test for stream errors.

dpgeorge commented on extmod/moddeflate.c:
```diff
@@ -0,0 +1,411 @@
+STATIC mp_uint_t deflateio_read(mp_obj_t o_in, void *buf, mp_uint_t size, int *errcode) {
+    mp_obj_deflateio_t *self = MP_OBJ_TO_PTR(o_in);
+
+    if (!deflateio_init_read(self) || self->stream == MP_OBJ_NULL) {
```
These checks should be swapped.

jimmo replied:
Done.

jimmo commented on extmod/moddeflate.c:
```diff
@@ -0,0 +1,411 @@
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_stream, MP_ARG_REQUIRED | MP_ARG_OBJ, {.u_obj = MP_OBJ_NULL} },
+        { MP_QSTR_format, MP_ARG_INT, {.u_int = DEFLATEIO_FORMAT_NONE} },
+        { MP_QSTR_wbits, MP_ARG_INT, {.u_int = 0} },
```
> Edit: or does `wbits=0` mean something special?

Yes.

jimmo committed:
examples/natmod/deflate: Add deflate as a dynamic native module.

This replaces the previous zlib version.

This work was funded through GitHub Sponsors.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
Added `deflate` as a dynamic native module example (to replace the old `zlib` version).

jimmo commented:
Made compression enabled at the "full" level by default, and updated docs to match.

dpgeorge commented on docs/library/deflate.rst:
```diff
@@ -0,0 +1,176 @@
+Classes
+-------
+
+.. class:: DeflateIO(stream, format=AUTO, wbits=0, *, close=False, /)
```
I don't think the `*` should be there.

dpgeorge commented on extmod/moddeflate.c:
```diff
@@ -0,0 +1,403 @@
+        // -----CMF------  ----------FLG---------------
+        // CINFO(5) CM(3)  FLEVEL(2) FDICT(1) FCHECK(5)
+        uint8_t buf[] = { 0x08, 0x80 };     // CM=2 (deflate), FLEVEL=2 (default), FDICT=0 (no dictionary)
+        buf[0] |= MAX(wbits - 8, 1) << 4;     // base-2 logarithm of the LZ77 window size, minus eight.
```
maybe line up the comments here?

dpgeorge commented on ports/nrf/modules/os/microbitfs.c:
```diff
@@ -609,7 +603,7 @@ STATIC const mp_map_elem_t os_mbfs_file_locals_dict_table[] = {
     { MP_OBJ_NEW_QSTR(MP_QSTR_name), (mp_obj_t)&os_mbfs_file_name_obj },
     { MP_ROM_QSTR(MP_QSTR___enter__), (mp_obj_t)&mp_identity_obj },
-    { MP_ROM_QSTR(MP_QSTR___exit__), (mp_obj_t)&os_mbfs_file___exit___obj },
+    { MP_ROM_QSTR(MP_QSTR___exit__), (mp_obj_t)&mp_stream___exit___obj },
```
I don't think this will work because this type doesn't implement ioctl.

dpgeorge commented on lib/uzlib/uzlib.h:
```diff
@@ -92,7 +84,8 @@ struct uzlib_uncomp {
        besides returning the next byte, it may also update source and
        source_limit fields, thus allowing for buffered operation. */
-    int (*source_read_cb)(struct uzlib_uncomp *uncomp);
+    void *source_read_data;
```
I'm not sure why this new `source_read_data` member is needed?

dpgeorge replied:
It's needed and used.

jimmo commented:
Updated to address comments. Added `AUTO` as a module constant.

dpgeorge commented:
Now merged!  Thanks @jimmo for all the hard work on this.
